### PR TITLE
Baseurl in uistate

### PR DESCRIPTION
**Pull request by jaycee**

### CODE REVIEW

**Comment by jaycee** [comment]
QA:

First confirm that the gui operates normally as you click between views.

Then, modify app/config-debug.js and set something (e.g. '/foo') as the baseUrl. Go to /foo on the gui and make sure it comes up in its default, no loaded view form. Then click around and make sure all the views work, all prefixed with /foo.

File: app/subapps/browser/browser.js [hunk 1]
```diff
@@ -176,10 +176,16 @@ YUI.add('subapp-browser', function(Y) {
...
+      var baseUrl = '';
+      if (cfg && cfg.baseUrl) {
+        baseUrl = cfg.baseUrl;
+      }
+
       this.state = new models.UIState({
         // Disallow routing to inspectors if we are in sandbox mode; the
         // model to be inspected will not be available.
         allowInspector: !cfg.sandbox,
+        baseUrl: baseUrl,
```
Comment (hatched) [inline app/subapps/browser/browser.js]: `baseUrl = cfg.baseUrl || '',`

File: app/subapps/browser/browser.js [hunk 2]
```diff
@@ -176,10 +176,16 @@ YUI.add('subapp-browser', function(Y) {
       // charms are selected.
       this._cache = new Y.juju.BrowserCache();
 
+      var baseUrl = '';
```
Comment (hatched) [inline app/subapps/browser/browser.js]: You can remove this block if you use the comment below

File: app/app.js [hunk 3]
```diff
@@ -650,11 +650,16 @@ YUI.add('juju-gui', function(Y) {
           ['add', 'remove', '*:change'],
           this.on_database_changed, this);
 
+      var baseUrl = '';
```
Comment (hatched) [inline app/app.js]:
```JavaScript
if (window.juju_config && window.juju_config.baseUrl) {
  cfg.baseUrl = window.juju_config.baseUrl;
}
```

File: app/app.js [hunk 4]
```diff
@@ -650,11 +650,16 @@ YUI.add('juju-gui', function(Y) {
...
+      if (window.juju_config && window.juju_config.baseUrl) {
+        baseUrl = window.juju_config.baseUrl;
+      }
       // Share the store instance with subapps.
       cfg.charmstore = this.get('charmstore');
       cfg.envSeries = this.getEnvDefaultSeries.bind(this);
       cfg.env = this.env;
       cfg.ecs = this.env.ecs;
+      cfg.baseUrl = baseUrl;
```
Comment (hatched) [inline app/app.js]: remove

Comment (jaycee) [inline app/app.js]: Isn't it better to set it to '' then to leave it undefined?

File: app/views/state.js [hunk 5]
```diff
@@ -320,6 +321,7 @@ YUI.add('juju-app-state', function(Y) {
       var url = req.path,
           query = req.query,
           state = {};
+      url = url.replace(this.get('baseUrl'), '');
```
Comment (hatched) [inline app/views/state.js]: It's trivial but can you add a comment here plz - I really want to keep this as documented as possible :)

File: test/test_ui_state.js [hunk 6]
```diff
@@ -841,6 +841,24 @@ describe('UI State object', function() {
       });
     });
 
+    it('parses urls with a baseUrl', function() {
+      var saveStub = testUtils.makeStubMethod(state, 'saveState');
+      this._cleanups.push(saveStub.reset);
+      var oldBaseUrl = state.get('baseUrl');
```
Comment (hatched) [inline test/test_ui_state.js]: A new state instance is generated for every test so this shouldn't be necessary. If it is could you please investigate further.

File: test/test_ui_state.js [hunk 7]
```diff
@@ -1104,6 +1122,30 @@ describe('UI State object', function() {
       });
     });
 
+    it('can generate proper urls with a baseUrl', function() {
+      var oldBaseUrl = state.get('baseUrl');
```
Comment (hatched) [inline test/test_ui_state.js]: See above

**Comment by hatched** [comment]
:+1: all comments trivials. 

No QA

File: app/app.js (hunk 4 above)
Comment (hatched) [inline app/app.js]: In my review of browser.js I modified it to set the default instead.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2645/
Test PASSed.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2647/
Test FAILed.

**Comment by jaycee** [comment]
Passing test run here: http://ci.jujugui.org:8080/job/juju-gui/2649/

**Comment by kadams54** [comment]
Ran into one issue during QA but the same problem shows up on develop, so it's not a problem with this PR. I'll raise it separate from this; QA is OK and :+1: on the code.

**Comment by jaycee** [comment]
Thanks all.

:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/2650/
Test PASSed.

**Pull request merged**